File: kernel/src/main/kotlin/ninja/blacknet/transaction/Burn.kt
/*
 * Copyright (c) 2018-2024 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.transaction

import kotlinx.serialization.Serializable
import ninja.blacknet.core.*
import ninja.blacknet.crypto.Hash
import ninja.blacknet.serialization.ByteArraySerializer
import ninja.blacknet.serialization.LongSerializer

@Serializable
class Burn(
        @Serializable(with = LongSerializer::class)
        val amount: Long,
        @Serializable(with = ByteArraySerializer::class)
        val message: ByteArray
) : TxData {
    override fun processCoinImpl(tx: Transaction, hash: Hash, dataIndex: Int, coinTx: CoinTx): Status {
        if (amount == 0L) {
            return Invalid("Invalid amount")
        }
        val account = coinTx.getAccount(tx.from)!!
        val status = account.credit(amount)
        if (status != Accepted) {
            return status
        }
        coinTx.setAccount(tx.from, account)
        coinTx.addSupply(-amount)
        return Accepted
    }
}

File: kernel/src/main/rust/transaction/burn.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::amount::Amount;
use crate::blake2b::Hash;
use crate::error::{Error, Result};
use crate::transaction::{CoinTx, Transaction, TxData};
use alloc::borrow::ToOwned;
use alloc::boxed::Box;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
pub struct Burn {
    amount: Amount,
    message: Box<[u8]>,
}

impl TxData for Burn {
    fn process_impl(
        &self,
        tx: Transaction,
        _hash: Hash,
        _data_index: u32,
        coin_tx: impl CoinTx,
    ) -> Result<()> {
        if self.amount == Amount::ZERO {
            return Err(Error::Invalid("Invalid amount".to_owned()));
        }
        let mut account = coin_tx.get_account(tx.from)?;
        account.credit(self.amount)?;
        coin_tx.set_account(tx.from, account);
        coin_tx.sub_supply(self.amount);
        Ok(())
    }
}
